includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php

```diff
@@ -133,7 +133,9 @@ public function maybe_init_subscriptions() {
 		add_filter( 'woocommerce_email_classes', [ $this, 'add_emails' ], 20 );
 		add_filter( 'woocommerce_available_payment_gateways', [ $this, 'prepare_order_pay_page' ] );
 
-		add_action( 'woocommerce_scheduled_subscription_payment_' . $this->id, [ $this, 'scheduled_subscription_payment' ], 10, 2 );
+		if ( ! has_action( 'woocommerce_scheduled_subscription_payment_' . $this->id ) ) {
```

Comment:
Having a safety check here to prevent multiple renewals seems really sensible one way or another.

That being said, how do you all feel about the current state of the UPE implementation that caused the extra call in the first place. Do we also need to fix the originating issue to bring this to a satisfactory fix?

Reply:
I'll defer to @FangedParakeet and others that are more familiar with the recent UPE work, but I think to root out the originating issue might be a large / regression prone task. My understanding is that we need to modify how / when we're calling the parent gateway (`WC_Payment_Gateway_WCPay`) constructor as it is not getting the UPE payment method IDs to properly name hooks, etc. However, this change would only really apply to Split UPE as I don't believe non-UPE and legacy UPE rename the methods at all. For a rough idea of what that might look like, it was [initially looked into here](https://github.com/Automattic/woocommerce-payments/pull/5530/commits/889b224032ff615d66715f21421c99d3a3b6d6ad), but in a quick test today that seemed to cause some issues in looking up the subscription (though I didn't put much time into it beyond that).

Reply:
@FangedParakeet's update makes this a much more thorough solution by avoiding the multiple calls to `maybe_init_subscriptions`. That change has now been rolled into this PR as well for a two pronged solution.

Reply:
If we check `has_action` before running this hook, how will this work if someone else also used this hook?

For example, if another plugin uses this action first:

```
add_action( 'woocommerce_scheduled_subscription_payment_' . $id, function() {} );
```

Then our `scheduled_subscription_payment` would not execute.

Reply:
I'm not saying this is impossible, but someone else would have to register the action `woocommerce_scheduled_subscription_payment_woocommerce_payments` in order to overwrite our own hook here. From our perspective, our goal is to ensure that this hook only fires once for any WCPay Subscription payment process. Ensuring that `maybe_init_subscriptions` only fires once, by calling it in a single place in `WC_Payments::init()` should also accomplish this, but this check is just to be twice as certain. 

Let us know if you think this is redundant or if there is a better way this could be handled!

Reply:
> If we check has_action before running this hook, how will this work if someone else also used this hook?

I think this is a good point and introducing that risk is something we should keep in mind.

While the chance is super low, it is possible. What makes it even more complicated is that if this were to happen, it would be pretty hard to discover that is the cause.